packages/tpetra/core/inout/MatrixMarket_TpetraNew.hpp

```diff
@@ -0,0 +1,986 @@
+  }
+
+  //  Add diagonal entries if they are required.
+  //  Check to make sure they are all here; add them if they are not.
+  if (requireDiagonal) {
+    if (dist->DistType() == MMFile) {
+      // All diagonal entries should be present in the file; we cannot create
+      // them for file-based data distributions.  
+      // Do an error check to ensure they are all there.
+      size_t localss = diagset.size();
+      size_t globalss;
+      Teuchos::reduceAll<int, size_t>(*comm, Teuchos::REDUCE_SUM, 1,
+                                      &localss, &globalss);
+      TEUCHOS_TEST_FOR_EXCEPTION(globalss != nRow, std::logic_error,
+        "File-based nonzero distribution requires all diagonal "
+         << "entries to be present in the file.  \n" 
+         << "Number of diagonal entries in file = " << globalss << "\n"
+         << "Number of matrix rows = " << nRow << "\n");
+    }
+    else {
+      for (global_ordinal_type i = 0;
+                               i < static_cast<global_ordinal_type>(nRow); i++) 
+      {
+        if (dist->Mine(i,i)) {
+          if (diagset.find(i) == diagset.end()) {
+            nzindex_t idx = std::make_pair(i,i);
+            localNZ[idx] = 0;   
+          }
+        }
+      }
+    }
+  }
+  // Done with diagset; free its memory
+  std::set<global_ordinal_type>().swap(diagset);
+
+  if (useTimers) 
+    timer = Teuchos::null;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+//////////////////  BINARY FILE READER & RESPECTIVE FORMAT ///////////////////
+//        (This format is used by the upcoming readBinary function)         //
+//////////////////////////////////////////////////////////////////////////////
+// 
+// File format:
+//   #vertices #edges src_1 dst_1 src_2 dst_2 ... src_#edges dst_#edges
+// 
+// Types:
+//   #edges:            unsigned long long
+//   everything else:   unsigned int
+//
+// Base of indexing: 
+//   1
+// 
+//////////////////////////////////////////////////////////////////////////////
+//
+// A code example that converts MatrixMarket format into this format:
+//
+//   typedef unsigned int ord_type;
+//   typedef unsigned long long size_type;
+//
+//   std::string line;
+//   std::ifstream in(infilename);
+//   std::ofstream out(outfilename,  std::ios::out | std::ios::binary);
+//
+//  ord_type nv, dummy, edge[2];
+//  size_type ne;
+//
+//   do
+//     std::getline (in, line);
+//   while(line[0] == '%');
+//
+//   std::stringstream ss(line);
+//   ss >> nv >> dummy >> ne;
+//   out.write((char *)&nv, sizeof(ord_type));
+//   out.write((char *)&ne, sizeof(size_type));
+//
+//   while(std::getline(in, line)) {
+//     std::stringstream ss2(line);
+//     ss2 >> edge[0] >> edge[1];
+//     out.write((char *)edge, sizeof(ord_type)*2);
+//
+//    }
+//   in.close();
+//   out.close();
+//
+//////////////////////////////////////////////////////////////////////////////
```

Comment:
@jhux2 @kddevin These comment lines describe the binary format and a code snippet to convert a MM file into this format. It only supports unweighted graphs for now. So, there are two basic differences between MM and this format:
1) Header information consists of numVertices and numEdges, so this format can not support a rectangular matrix.
2) Values are not supported either, that is, each nonzero entry only consists of the source and destination vertices (i.e., rowId and colId).

Reply:
@seheracer So does this mean that the binary format only supports graphs right now, not numeric matrices?  Just want to be sure I understand.

Reply:
Correct. Numeric matrices are not supported by this format yet.